ecommerce/views/v0/__init__.py

```diff
@@ -136,3 +169,500 @@ def create(self, request, *args, **kwargs):
         )
+
+
+class CheckoutApiViewSet(ViewSet):
+    authentication_classes = (SessionAuthentication, TokenAuthentication)
+    permission_classes = (IsAuthenticated,)
+
+    @action(
+        detail=False,
+        methods=["post"],
+        name="Redeem Discount",
+        url_name="redeem_discount",
+    )
+    def redeem_discount(self, request):
+        """
+        API call to redeem a discount. Discounts are attached to the basket so
+        they can be attached at any time in the checkout process. Later on,
+        they'll be converted to attach to the resulting Order (as Baskets are
+        ephemeral).
+
+        POST Args:
+            - discount (int): Discount ID to apply
+
+        Returns:
+            - Success message on success
+            - HTTP 406 if there's no basket yet
+            - HTTP 404 if the discount isn't found
+        """
+        try:
+            basket = Basket.objects.filter(user=request.user).get()
+        except ObjectDoesNotExist:
+            return Response("No basket", status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        try:
+            discount = Discount.objects.get(pk=request.data["discount"])
+        except ObjectDoesNotExist:
+            return Response("Discount not found.", status=status.HTTP_404_NOT_FOUND)
+
+        basket_discount = BasketDiscount(
+            redemption_date=now_in_utc(),
+            redeemed_by=request.user,
+            redeemed_discount=discount,
+            redeemed_basket=basket,
+        )
+        basket_discount.save()
+
+        # This should maybe recalculate the pricing for stuff in the basket?
+
+        return Response({"message": "Discount applied", "id": basket_discount.id})
+
+    @action(
+        detail=False, methods=["post"], name="Start Checkout", url_name="start_checkout"
+    )
+    def start_checkout(self, request):
+        """
+        API call to start the checkout process. This assembles the basket items
+        into an Order with Lines for each item, applies the attached basket
+        discounts, and then calls the payment gateway to prepare for payment.
+
+        Returns:
+            - JSON payload from the ol-django payment gateway app. The payment
+              gateway returns data necessary to construct a form that will
+              ultimately POST to the actual payment processor.
+        """
+        try:
+            basket = Basket.objects.filter(user=request.user).get()
+        except ObjectDoesNotExist:
+            return Response("No basket", status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        discounts = BasketDiscount.objects.filter(redeemed_basket=basket).all()
+
+        total = 0
+
+        order = PendingOrder(total_price_paid=0, purchaser=request.user)
+        order.save()
+
+        x_forwarded_for = request.META.get("HTTP_X_FORWARDED_FOR")
+        if x_forwarded_for:
+            ip = x_forwarded_for.split(",")[0]
+        else:
+            ip = request.META.get("REMOTE_ADDR")
```

Comment:
Would be worth pulling in [`django-ipware`](https://pypi.org/project/django-ipware/) to handle determination of the IP address.